high_performance_computing/hpc_mpi/01_introduction.md

```diff
@@ -0,0 +1,399 @@
+
+When the data is structured in a regular way, such as when simulating atoms in a crystal, it makes sense to divide the space into domains. 
+Each rank will handle the simulation within its own domain.
+
+![Data points divided into four ranks](fig/domaindecomposition.png)
+
+Many algorithms involve multiplying very large matrices.
+These include finite element methods for computational field theories as well as training and applying neural networks.
+The most common parallel algorithm for matrix multiplication divides the input matrices into smaller submatrices and composes the result from multiplications of the submatrices.
+If there are four ranks, the matrix is divided into four submatrices.
+
+$$ A = \left[ \begin{array}{cc}A_{11} & A_{12} \\ A_{21} & A_{22}\end{array} \right] $$
+$$ B = \left[ \begin{array}{cc}B_{11} & B_{12} \\ B_{21} & B_{22}\end{array} \right] $$
+
+$$ A \cdot B = \left[ \begin{array}{cc}A_{11} \cdot B_{11} + A_{12} \cdot B_{21} & A_{11} \cdot B_{12} + A_{12} \cdot B_{22} \\ A_{21} \cdot B_{11} + A_{22} \cdot B_{21} & A_{21} \cdot B_{12} + A_{22} \cdot B_{22}\end{array} \right]$$
+
+If the number of ranks is higher, each rank needs data from one row and one column to complete
+its operation.
+::::
+
+::::callout
+#### Load Balancing
+Even if the data is structured in a regular way and the domain is decomposed such that each core can take charge of roughly equal amounts of the sub-domain, the work that each core has to do may not
+be equal. 
+For example, in weather forecasting, the 3D spatial domain can be decomposed in an equal portion. But when the sun moves across the domain, the amount of work is different in that domain since more
+complicated chemistry/physics is happening in that domain.
+Balancing this type of loads is a difficult problem and requires a careful thought before designing a parallel algorithm.
+::::
+
+:::::challenge{id=serial-and-parallel, title="Serial and Parallel Regions"}
+Identify the serial and parallel regions in the following algorithm:
+
+```
+vector_1[0] = 1;
+vector_1[1] = 1;
+for i in 2 ... 1000
+  vector_1[i] = vector_1[i-1] + vector_1[i-2];
+
+  for i in 0 ... 1000
+    vector_2[i] = i;
+
+  for i in 0 ... 1000
+    vector_3[i] = vector_2[i] + vector_1[i];
+    print("The sum of the vectors is.", vector_3[i]);
+```
+
+::::solution
+```
+serial   | vector_0[0] = 1;
+         | vector_1[1] = 1;
+         | for i in 2 ... 1000
+         |   vector_1[i] = vector_1[i-1] + vector_1[i-2];
+
+parallel | for i in 0 ... 1000
+         |   vector_2[i] = i;
+
+parallel | for i in 0 ... 1000
+         |   vector_3[i] = vector_2[i] + vector_1[i];
+         |   print("The sum of the vectors is.", vector_3[i]);
+```
+The first and the second loop could also run at the same time.
+
+In the first loop, every iteration depends on data from the previous two.
+In the second two loops, nothing in a step depends on any of the other steps, and therefore can be parallelised.
+::::
+:::::

```

Comment:
Ah, yeah - that's actually identical in our DiRAC OpenMP and MPI courses, so it would make sense to break it out to a shared one here (`hpc_parallel_intro`).